### PR TITLE
Allow chaining expressions with Closure objects

### PR DESCRIPTION
**Pull request by jf2048**
I had to change `ClosureExpression::with_closure` to expand out the generic parameters, it could be changed back depending on this: https://github.com/rust-lang/rust/issues/83701

Also includes some docs fixes.

### CODE REVIEW

**Comment by SeaDve** [comment]
I think `chain_gclosure` may be clearer

**Comment by jf2048** [comment]
That may not be straightforward as the macro was changed to just `closure!`. Maybe it could be `chain_with_closure`, but that also seems confusing.

**Comment by bilelmoussaoui** [comment]
Maybe rename the first `chain_closure` to `chain_callback` & keep this one as `chain_closure`?

**Comment by SeaDve** [comment]
I just realized that it was named `chain_closure` before since it chains a `ClosureExpression`, but something like `chain_closure_with_closure` is a weird name

**Comment by bilelmoussaoui** [comment]
> I just realized that it was named `chain_closure` before since it chains a `ClosureExpression`, but something like `chain_closure_with_closure` is a weird name

Well the ClosureExpression should only take a glib::Closure but we provide a helper constructor that takes a callback. So it's fine to not name it  `chain_closure` if it's documented properly. The double `chain_closure_with_closure` is definitely not great

**Comment by SeaDve** [comment]
I see, maybe `chain_closure` and `chain_closure_with_callback` ?

**Comment by bilelmoussaoui** [comment]
> I see, maybe `chain_closure` and `chain_closure_with_callback` ?

No strong opinions here, @sdroege ?:)

**Comment by sdroege** [comment]
Seems fine either way. Ask your favorite random number generator and go with that :)

**Comment by bilelmoussaoui** [comment]
> I see, maybe `chain_closure` and `chain_closure_with_callback` ?

Let's go with this one then

**Comment by SeaDve** [comment]
semi related to pr, maybe the ClosureExpression could also ne renamed to `new` and `with_callback` for coherence

File: gdk4/src/event.rs [hunk 1]
```diff
@@ -281,7 +281,7 @@ impl AsRef<Event> for Event {
 }
 
 // rustdoc-stripper-ignore-next
-/// A common trait implemented by the various [`gdk::Event`] types.
+/// A common trait implemented by the various [`gdk::Event`](crate::Event) types.
```
Comment (bilelmoussaoui) [inline gdk4/src/event.rs:284]: I fixed those as part https://github.com/gtk-rs/gtk4-rs/pull/719/ btw :-)

**Comment by jf2048** [comment]
This is updated with the name changes, but it's a bit awkward with the recent addition of `RustClosure` because now it always has to be `closure!( ... ).into()`.

File: gtk4/src/closure_expression.rs [hunk 2]
```diff
@@ -23,16 +23,13 @@ define_expression!(
 
 impl ClosureExpression {
     #[doc(alias = "gtk_closure_expression_new")]
-    pub fn new<F, R>(params: impl IntoIterator<Item = impl AsRef<Expression>>, callback: F) -> Self
+    pub fn new<R, I, E>(params: I, closure: glib::Closure) -> Self
```
Comment (bilelmoussaoui) [inline gtk4/src/closure_expression.rs:26]: Make it take a RustClosure ?

**Comment by bilelmoussaoui** [comment]
other than the cargo fmt, lgtm

File: gtk4/src/closure_expression.rs (hunk 2 above)
Comment (sdroege) [reply gtk4/src/closure_expression.rs:26]: It probably has to for type-safety reasons anyway.

Comment (jf2048) [reply gtk4/src/closure_expression.rs]: I changed it to `RustClosure`, if someone wants to add another function that takes just `Closure` it should probably be unsafe anyway.

**Comment by bilelmoussaoui** [comment]
Thanks!

**Pull request merged**